### PR TITLE
Add ability to connect to Redis with client SSL/TLS certificates.

### PR DESCRIPTION
**Pull request by mdayaram**
SSL connections were supported before, but there was no way to provide
client certificates.  This changeset introduces REDBEAT_REDIS_USE_SSL to
provide further SSL options, with a fallback to BROKER_USE_SSL,
mimicking the behavior we have we REDBEAT_REDIS_URL and BROKER_URL.

Small mention in #99 though they didn't need client certs there.

### CODE REVIEW

**Comment by mdayaram** [comment]
@sibson let me know if there's anything I can do to help move this along, have to update/change anything.

File: redbeat/schedulers.py [hunk 1]
```diff
@@ -246,7 +253,7 @@ def due_at(self):
 
     @property
     def key(self):
-        return self.app.redbeat_conf.key_prefix + self.name
+        return self.app.redbeat_conf.key_prefix + str(self.name)
```
Comment (sibson) [inline redbeat/schedulers.py]: Why ```str()``` here? It seems unrelated to the PR.

File: tests/test_scheduler.py [hunk 2]
```diff
@@ -223,3 +224,98 @@ def setup(self): # celery3
...
+    config_dict = {
+            'REDBEAT_KEY_PREFIX': 'rb-tests:',
+            'REDBEAT_REDIS_URL': 'rediss://redishost:26379/0',
+            'REDBEAT_REDIS_OPTIONS': {
+                'password': '123',
+                'service_name': 'master',
+                'socket_timeout': 0.1,
+            },
+            'CELERY_RESULT_BACKEND' : 'redis-sentinel://redis-sentinel:26379/1',
+            'REDBEAT_REDIS_USE_SSL': {
```
Comment (sibson) [inline tests/test_scheduler.py]: Why are the ```ssl_*``` options in a new config key rather than folding them into ```REDBEAT_REDIS_OPTIONS```?

File: redbeat/schedulers.py (hunk 1 above)
Comment (mdayaram) [reply redbeat/schedulers.py]: Ah, sorry, yeah, i forgot to mention that I kept getting type errors in this line regarding bytes vs strings here.  I narrowed it down to name being in bytes instead of strings, which fixed the issue.  I tried to trace where `name` was being set, but had a hard time figuring out.  If you know where that is, I could investigate further and find out why it's being set as bytes instead of strings for my usecase.

File: tests/test_scheduler.py (hunk 2 above)
Comment (mdayaram) [reply tests/test_scheduler.py]: Yeah, I thought about that but then figured I wanted to mirror the options for the Broker.  From the looks at the README and broker options, `BROKER_URL <=> REDBEAT_REDIS_URL`, `BROKER_TRANSPORT_OPTIONS <=> REDBEAT_REDIS_OPTIONS` -- so I figured having `BROKER_USE_SSL <=> REDBEAT_REDIS_USE_SSL` was the natural way to go.  Happy to change it if you think adding them to `REDBEAT_REDIS_OPTIONS` is better.

File: redbeat/schedulers.py (hunk 1 above)
Comment (sibson) [reply redbeat/schedulers.py]: name is either coming from [configuration](https://github.com/sibson/redbeat/blob/master/redbeat/schedulers.py#L365) or [existing in redis](https://github.com/sibson/redbeat/blob/master/redbeat/schedulers.py#L401).  Or it  could be creating Entries programmatically somewhere in your code.   If there is a problem with RedBeat handling name I'd like to see that fixed in a separate PR.

File: tests/test_scheduler.py (hunk 2 above)
Comment (sibson) [reply tests/test_scheduler.py]: Thanks I see your logic now, and agree it follows the pattern established upstream

.  It might be worth evaluating if we actually want to support ```REDBEAT_REDIS_USE_SSL=True``` or not.  From the [celery docs](http://docs.celeryproject.org/en/latest/userguide/configuration.html#broker-use-ssl) that could allow users a false sense of security and forcing them to be explicit would avoid security mistakes.  What do you think?

File: redbeat/schedulers.py (hunk 1 above)
Comment (mdayaram) [reply redbeat/schedulers.py]: Thanks for the leads, I'll investigate that separately and open another PR when I find the culprit.  Thanks!

File: tests/test_scheduler.py (hunk 2 above)
Comment (mdayaram) [reply tests/test_scheduler.py]: Hmm, well, I'm relatively new to python, so not sure what the better convention here would be.  In our case, I made sure that if `REDBEAT_REDIS_USE_SSL` was equal to `True`, that we were verifying the certificate by having `ssl.CERT_REQUIRED` be in the default dictionary.  So in terms of security safety, I think we'd be covered.

I can see arguments for either case regarding conventions though, on one hand, setting the value to `True` is an anti-pattern since not _every_ library guarantees that it'll verify certificates and we could actively discourage this pattern by not allowing it, but on the other hand, allowing `True` but in a safe way makes this library more consistent with other libraries that are like it.

Since it's more of a convention issue, I'm happy to yield to whatever you think is best here.

File: redbeat/schedulers.py (hunk 1 above)
Comment (mdayaram) [reply redbeat/schedulers.py]: Removed this change from the latest update 👍

File: redbeat/schedulers.py [hunk 3]
```diff
@@ -116,6 +117,11 @@ def get_redis(app=None):
                                 password=redis_options.get('password'),
                                 decode_responses=True)
             connection = sentinel.master_for(redis_options.get('service_name', 'master'))
+        elif conf.redis_url.startswith('rediss'):
+            ssl_options = { 'ssl_cert_reqs': ssl.CERT_REQUIRED }
```
Comment (sibson) [inline redbeat/schedulers.py]: I think there is a valid use cases where a snakeoil cert is being used along with a ```rediss://``` URL, which means we can't blindly enforce CERT_REQUIRED. As per our other conversation, this feels wrong from a UX perspective but is in line with existing behaviour in other packages in the ecosystem.

File: tests/test_scheduler.py (hunk 2 above)
Comment (sibson) [reply tests/test_scheduler.py]: As per my comment on https://github.com/sibson/redbeat/pull/110/files#diff-3306d7d435bbd01ef84d83cae122c8cfR121, it seems like we won't be able to protect users from themselves by forcing CERT_REQUIRED.  It seems the best we can hope to do is document the best practice.

File: redbeat/schedulers.py (hunk 3 above)
Comment (mdayaram) [reply redbeat/schedulers.py]: Ah, good point.  In that case, should I could raise an exception if the value of `redbeat_redis_use_ssl` is not a dictionary?

File: tests/test_scheduler.py (hunk 2 above)
Comment (mdayaram) [reply tests/test_scheduler.py]: Hmmm, sure, I'll add a bit in the README similar to what the celery docs have for the broker.

File: redbeat/schedulers.py (hunk 3 above)
Comment (mdayaram) [reply redbeat/schedulers.py]: @sibson What do you think? I think, given that the user can still override the value to `CERT_NONE` in the case they're using snakeoil certs, that the current implementation should be fine.  Happy to change it to something else if you think there's a better way.

Comment (sibson) [reply redbeat/schedulers.py]: Thanks for pointing that out.  It's pretty reasonable stance to take.  My hesitation would be that it's a breaking backwards incompatible change. Being as it should fail early and spectacularly, I think we can include it as is with a version bump and clearly documented.  I'll merge as is once I check if we need a release of current master before doing the version bump.

Comment (mdayaram) [reply redbeat/schedulers.py]: Ah, right, backwards compatibility!  Depending on how much work you want to do, we could release a version of this change that's backwards compatible but unsafe, having it log a warning saying the behavior will change in the major version, and then merge this PR as is on the major version.  Just waiting for the major version is fine too though.

Either way, thanks for taking a look and helping me flesh this out and getting it in =)

**Comment by mdayaram** [comment]
Hi @sibson -- just wondering about the progress of merging this ticket in, anything I can do to expedite? Happy to create a backwards-compatible PR to merge into the current version if that would help.

**Comment by mdayaram** [comment]
ping -- hey, just wanted to get a status update -- anything I can do to help get this in?

File: redbeat/schedulers.py (hunk 3 above)
Comment (sibson) [reply redbeat/schedulers.py]: I'm tagging this for 2.0 release, once I clear the decks on a 1.0 release I'll merge this.

Comment (mdayaram) [reply redbeat/schedulers.py]: Cool, thanks for the update!

**Comment by mdayaram** [comment]
Hi @sibson -- did this get reverted for good?

**Comment by sibson** [comment]
Hey, sorry I lost track of this when trying to fix tests.  I'd be happy to include this again.  I've dropped support for celery 3, so this may need a few small updates before it gets re-introduced.

**Comment by sibson** [comment]
Looking at this history, I created the branch to test reverting the change, but didn't merge the branch.  So you should be able to use certs now.

**Comment by mdayaram** [comment]
Awesome to hear! Do you know when the next release will be? I think the latest version in pypi is still one without this change.
